File: examples/iterator.js
function next(self: Iterator): number {
  return self.val += 1;
}

const iterator = { val: 0, next };

console.log("Returned: " + iterator.next(iterator) + '\n');
console.log("Returned: " + iterator.next(iterator) + '\n');
console.log("Returned: " + iterator.next(iterator) + '\n');
console.log("Returned: " + iterator.next(iterator) + '\n');
console.log("Returned: " + iterator.next(iterator) + '\n');
console.log("Returned: " + iterator.next(iterator) + '\n');

File: src/runtime/core/iterator.rs
use crate::tokens::{AmvmType, AmvmTypeDefinition};

pub fn amvm_iterator_type() -> AmvmTypeDefinition {
    #[allow(non_snake_case)]
    let T: Box<AmvmType> = Box::new(AmvmType::Named(Box::from("T")));

    AmvmTypeDefinition::Struct {
        generics: vec![("T".into(), None)],
        fields: vec![(
            "next".into(),
            AmvmType::Fun(vec![AmvmType::Named("Iterator".into())], T),
        )],
    }
}
